xmtp_proto/src/types/group_message.rs

```diff
@@ -0,0 +1,71 @@
+use super::{Cursor, GroupId};
+use crate::ConversionError;
+use chrono::Local;
+use derive_builder::Builder;
+use openmls::prelude::ContentType;
+
+/// A GroupMessage from the network
+#[derive(Clone, Builder)]
+#[builder(setter(into), build_fn(error = "ConversionError"))]
+pub struct GroupMessage {
+    /// Cursor of this message
+    pub cursor: Cursor,
+    /// server timestamp indicating when this message was created
+    pub created_ns: chrono::DateTime<Local>,
+    /// GroupId of the message
+    pub group_id: GroupId,
+    // MLS Group Message
+    pub message: openmls::framing::ProtocolMessage,
+    /// Sender HMAC key
+    pub sender_hmac: Vec<u8>,
+    /// Whether this message should result in a push notification
+    pub should_push: bool,
+    /// Payload hash of the message
+    /// TODO: make payload hash constant array
+    pub payload_hash: Vec<u8>,
+}
+
+impl GroupMessage {
+    pub fn builder() -> GroupMessageBuilder {
+        GroupMessageBuilder::default()
+    }
+
+    pub fn is_commit(&self) -> bool {
+        self.message.content_type() == ContentType::Commit
+    }
+
+    pub fn timestamp(&self) -> i64 {
+        self.created_ns.timestamp_nanos_opt().expect("timestamp out of range for i64, are we in 2262 A.D?")
+    }
+
+    pub fn originator_id(&self) -> u32 {
+        self.cursor.originator_id
+    }
+
+    pub fn sequence_id(&self) -> u64 {
+        self.cursor.sequence_id
+    }
+}
+
+#[cfg(any(test, feature = "test-utils"))]
+mod test {
+    use chrono::DateTime;
+    use openmls::prelude::PublicMessage;
+    use xmtp_common::{rand_vec, Generate};
+
+    use super::*;
+
+    impl Generate for GroupMessage {
```

Comment:
a common pattern in tests is generating a type populated with dummy data. this PR introduces a Generate trait to simplify fixing tests and consolidate/remove more duplicated code